### ISSUE TITLE
Improve performance of graph analysis

### ISSUE THREAD

**Issue by alexandrebouchard** [UBC-Stat-ML/blangSDK#71]
Gets slow (from 5s to 5min) between 300k and 3M objects in the graph, currently roughly between 10k and 100k random variables. Probably L3/RAM transition. In process of checking if its linear after that. Update: actually seems non-linear at at least 2 points! see 32 min for 200k in data below. So there might be an algorithmic problem here. Or a bad hashcode (although java 8 should shield us from that).

In some cases you may be able to vectorize and that will be essentially constant provided the individual components are not accessed individually. e.g.: idd data. or large sparse field sampled using custom made samplers (then would need to somehow disable breaking matrix into constituents; e.g use skip dependency annotations).

Other cases fundamentally need the large graph. A couple of potential directions:

- Don't include in the graph things that are known to have only one parent! This include certain kinds of Constituent nodes, e.g. FieldNodes. I.e just a graph expension trick, no need to do large rewrite for that.
- The graph library we use is probably bloated (check), e.g. might rely on a bunch of hash table. We could rewrite the graph traversal and linearization.

**Comment by alexandrebouchard** [comment]
Data:

100
3793
4092
107.2 ms

1_000
37085
40084
437.2 ms

10_000
366421
396420
5.482 s

<L3-RAM cache frontier?> L3 is 6MB

100_000
3762181
4062180
20.31 min


===  after commit f4c09d64bfd7fe687558529c3f93fabbcc60a312

1_000
401.4 ms

5_000
1.766 s

10_000
280017
290016
4.787 s

100_000
2800017
2900016
6.493 min

200_000
31.94 min

300_000
running

**Comment by alexandrebouchard** [comment]
New test case:

```
package blang

model TestPerf {
  param List<Integer> test ?: {
    val l = new LinkedList<Integer>
    for (i : 0 .. 1_000_000)
      l.add(0)
    l
  }
  laws {}
}
```

10K - 209.7ms
100K - 1.9s
1M - 49s
10M - running out of memory for 10Gig

So at least one factor is that there is a bit bloating factor memory-wise: the 10M Integers linked list costs about 250MB to store but its analysis breaks the 10g ceiling. That's a 40x or more bloat. 

First guess for explanation for the bloat (seems incorrect): AbstractBaseGraph in jgrapht uses specifics, which in turns has LinkedHashMap<V, DirectedEdgeContainer<V, E>>,  and DirectedEdgeContainer has two sets in it... actually no seems fine, internally using ArrayUnenforcedSet

**Comment by alexandrebouchard** [comment]
Since the breaking point of this example

**Comment by alexandrebouchard** [comment]
```
package blang

model HMMPerf {
  param int size ?: 10
  random List<RealVar> chain ?: latentRealList(size)
  laws {
    chain.get(0) ~ Normal(0.0, 1.0)
    for (int i : 1 ..< size) {
      chain.get(i) | RealVar prev = chain.get(i-1) ~ Normal(prev, 1.0)
    }
  }
}
```

10 67.26ms
100 116.5ms
1_000 357.2ms
10_000 4.4s
100_000 3.1min (> 4x slow down)

**Comment by alexandrebouchard** [comment]
Comparing to iid model 

```
package blang

model HMMPerf {
  param int size ?: 100_000
  random List<RealVar> chain ?: latentRealList(size)
  laws {
    chain.get(0) ~ Normal(0.0, 1.0)
    for (int i : 1 ..< size) {
      chain.get(i) /* | RealVar prev = chain.get(i-1) */ ~ Normal(0.0, 1.0)
    }
  }
}
```

1_000 300ms
10_000 3.8
100_000 2.967min 

So not huge difference iid vs HMM

**Comment by alexandrebouchard** [comment]
Without laws

```
package blang

model HMMPerf {
  param int size ?: 10_000
  random List<RealVar> chain ?: latentRealList(size)
  laws {
//    chain.get(0) ~ Normal(0.0, 1.0)
//    for (int i : 1 ..< size) {
//      chain.get(i) /* | RealVar prev = chain.get(i-1) */ ~ Normal(0.0, 1.0)
//    }
  }
}
```

10k 2.6 s
100k 2.995min

**Comment by alexandrebouchard** [comment]
Ah ah! Found likely culprit: Array based implementation of set of neighbors

in the above changing ``latentRealList(size)`` to ``new LinkedList(latentRealList(size))`` brings down the cost from 2.9 minutes to 4s!

Fix: changing the implementations details in jgrapht's Specifics

**Comment by alexandrebouchard** [comment]
Unfortunately problem still persists with Hash based Specifics.

**Comment by alexandrebouchard** [comment]
But we are now down to a very simple problematic example:

```
package blang

model HMMPerf {
  param int size ?: 100_000
  random int [] chain ?: ArrayLiterals::newIntArrayOfSize(size)
  laws {
  }
}
```

**Comment by alexandrebouchard** [comment]
Plot thickens.. this is slow too:

```
val watch = Stopwatch.createStarted
    val DirectedGraph<Integer, Pair<Integer, Integer>> graph = GraphUtils::newDirectedGraph
    val size = 100_000
    graph.addVertex(-1)
    for (i : 0 ..< size) {
      graph.addVertex(i)
      graph.addEdge(i-1, i)
    }
    println(watch)
```

2.341 min (but with profiling)

**Comment by alexandrebouchard** [comment]
Found some crummy things in JGraphT's AbstractBaseGraph. Performance seems to be tuned on O(degree) which can be bad for star shaped graphs. 

But does not explain the above with (i-1, i) ...

```
public E getEdge(V sourceVertex, V targetVertex)
        {
            if (containsVertex(sourceVertex)
                && containsVertex(targetVertex))
            {
                DirectedEdgeContainer<V, E> ec = getEdgeContainer(sourceVertex);

                Iterator<E> iter = ec.outgoing.iterator();

                while (iter.hasNext()) {
                    E e = iter.next();

                    if (getEdgeTarget(e).equals(targetVertex)) {
                        return e;
                    }
                }
            }

            return null;
        }
```

**Comment by alexandrebouchard** [comment]
Ah! ah! the (i-1, i) was due to poor hash code: now this is super fast:

```
    graph.addVertex(-1  * 123141)
    for (i : 0 ..< size) {
      val subw = Stopwatch.createStarted
      graph.addVertex(i * 123141)
      
      graph.addEdge((i-1) * 123141, i * 123141)
      if (i % 1000 == 0.0)
        println("" + i + " " + subw.elapsed(TimeUnit.MICROSECONDS))
    }
    println(watch)

```

**Comment by alexandrebouchard** [comment]
Fixed in bayonet 4.1.6

Also documented problem in https://github.com/jgrapht/jgrapht/issues/812
